File: Leetcode/Daily/06-2025/C++/23.cpp
// Problem: 2081. Sum of k-Mirror Numbers (Hard)
// Link: https://leetcode.com/problems/sum-of-k-mirror-numbers

class Solution {
public:
    bool isValid(string num) {
        int n = num.size();
        
        for (int i = 0; i < n / 2; i++) {
            if (num[i] != num[n - i - 1]) return false;
        }

        return true;
    }

    long long baseTo10(string num, int base) {
        long long ans = 0, pow = 1;
        
        for (int i = num.size() -1; i >= 0; i--) {
            ans += (num[i]-'0') * pow;
            pow *= base;
        }
        
        return ans;
    }

    string next(string curr, int base) {
        int n = curr.size();

        if (curr == string(n, base-1 + '0')) {
            return "1" + string(n-1, '0') + "1";
        }

        string ans = curr;

        for (int i = n/2; i < n; i++) {
            if (curr[i] - '0' < base - 1) {
                ans[i] = ans[n-i-1] = curr[i] + 1;
                break;
            } else {
                ans[i] = ans[n-i-1] = '0';
            }
        }

        return ans;

    }

    long long kMirror(int k, int n) {
        long long ans = 0;
        string curr = "1";

        while (n > 0) {
            long long dec = baseTo10(curr, k);
            if (isValid(to_string(dec))) {
                ans += dec;
                n--;
            }

            curr = next(curr, k);
        }

        return ans;
    }
};

File: Leetcode/Daily/06-2025/TypeScript/23.ts
// Problem: 2081. Sum of k-Mirror Numbers (Hard)
// Link: https://leetcode.com/problems/sum-of-k-mirror-numbers

function kMirror(k: number, n: number): number {
    function isValid(num: string) : boolean {
        let n = num.length;
        
        for (let i = 0; i < n / 2; i++) {
            if (num[i] != num[n - i - 1]) return false;
        }

        return true;
    }

    function baseTo10(num: string, base: number): number {
        let ans = 0, pow = 1;
        
        for (let i = num.length -1; i >= 0; i--) {
            ans += (num[i].charCodeAt(0) - '0'.charCodeAt(0)) * pow;
            pow *= base;
        }
        
        return ans;
    }

    function next(curr: string, base: number): string {
        let n = curr.length;

        let maxDigit = String.fromCharCode(base - 1 + '0'.charCodeAt(0));
        if (curr === maxDigit.repeat(n)) {
            return '1' + '0'.repeat(n - 1) + '1';
        }

        let ans = curr.split('');

        for (let i = Math.floor(n / 2); i < n; i++) {
            if (ans[i] < maxDigit) {
                let newChar = String.fromCharCode(curr[i].charCodeAt(0) + 1);                
                ans[n-i-1] = newChar;
                ans[i] = newChar;

                break;
            } else {
                ans[i] = ans[n-i-1] = '0';
            }
        }

        return ans.join('');
    }

    let ans = 0;
    let curr = "1";

    while (n > 0) {
        let dec = baseTo10(curr, k);

        if (isValid(dec.toString())) {
            ans += dec;
            n--;
        }

        curr = next(curr, k);
    }

    return ans;
};
